xdsl/transforms/dead_code_elimination.py

```diff
@@ -3,19 +3,58 @@
 from xdsl.ir import Operation
 from xdsl.passes import ModulePass
 from xdsl.pattern_rewriter import PatternRewriter, PatternRewriteWalker, RewritePattern
-from xdsl.traits import IsTerminator, SymbolOpInterface, is_side_effect_free
+from xdsl.traits import (
+    IsTerminator,
+    MemoryEffect,
+    MemoryEffectKind,
+    RecursiveMemoryEffect,
+    SymbolOpInterface,
+)
 
 
 def is_trivially_dead(op: Operation):
-    # Check that operation is side-effect-free and unused
+    # Returns if the operation has no observable side-effect.
     return (
-        not op.get_trait(IsTerminator)
-        and not op.get_trait(SymbolOpInterface)
-        and is_side_effect_free(op)
-        and all(not result.uses for result in op.results)
+        all(not result.uses for result in op.results)
+        and (not op.get_trait(IsTerminator))
+        and (not op.get_trait(SymbolOpInterface))
+        and would_be_trivially_dead(op)
     )
 
 
+def would_be_trivially_dead(rootOp: Operation) -> bool:
+    effecting_ops = {rootOp}
```

Comment:
Please add docstring, and maybe change the name. Is it lifted from MLIR? I'm not sure what the "would be" means

Reply:
It is lifted from MLIR.

Reply:
Then a docstring would_be really helpful :)

Reply:
Done both!